### ISSUE TITLE
Unify terminology: error vs. check vs. rule

### ISSUE THREAD

**Issue by not-my-profile** [charliermarsh/ruff#1546]
The documentation currently uses all of them switching between them at random. For example in the `--help`:

* `--select`: Comma-separated list of **error** codes to enable (or ALL, to enable all **checks**)
* `--exclude` List of paths, used to exclude files and/or directories from **checks**
* `--explain` Explain a **rule**

The README also contains "lint rule" thrice, "lint check" once, "check code prefix" 7 times and "error code prefix" once.

It would be better to use consistent terminology to avoid confusing the user.

Prior art:

Linter | Term used
--|--
clippy | lint(s)
ESLint | rule(s)
pyflakes | check(s) (but not used often)
pycodestyle | check(s) & error code(s)
pydocstyle | error(s) & error code(s)

Some thoughts:

* If rules can be both errors or warnings as suggested in [#1256](https://www.github.com/charliermarsh/ruff/issues/1256), then we probably want to discontinue any uses of the term "error code".
* Ruff is a linter so lint(s) might be the most obvious choice semantically, however users are less likely to be familiar with the noun "lint" than with a more common noun like "rule" or "check".
* Some lint(s) perform several checks or enforce several rules. For example TID251 checks both import statements as well as attribute access, T100 checks for several imports, etc. With that in mind "lint(s)" might make the most sense.
* While it is not strictly necessary that our UI lingo matches the lingo we use within the codebase, I think for the sake of consistency and seamless onboarding of contributors we should aim for consistent terminology throughout the project (CLI, API, documentation). I do consider the consistency of the user-facing terminology to be most important.

**Comment by charliermarsh** [comment]
Yes, good idea.

My thoughts:

- For the reasons you describe, I also dislike "error" or "error code".
- We probably need two terms: one for the abstract rule (like "lines are too long"), and one for a rule that has been triggered. For example, we could have "rule" and "violation" (in the code, this is basically `CheckKind` and `Check`).
- We also need terms for what are currently called "check code" and "check code prefix".

I think my preference is "rule" and "violation" (which could be more specifically broken down into errors and warnings, both of which are kinds of violations).

**Comment by charliermarsh** [comment]
@andersk - Sorry to ping and please feel free to ignore if uninterested but you tend to have good and thoughtful opinions on this kind of stuff. Any preferences on the stabilization of this nomenclature?

**Comment by charliermarsh** [comment]
When we change this, we should also change the internal convention that "plugin" refers to any function that takes a `Checker` and adds a rule, where "check" refers to a pure function that returns a `Check`. It's super confusing.

**Comment by andersk** [comment]
Perhaps “diagnostic” might be less unfamiliar than “lint” and less judgmental than “error” or “violation”? Some other possibilities that came to mind are “defect” and “issue”.

I recently noticed the [SARIF specification](https://www.oasis-open.org/committees/sarif/), which seems to have “rules” and “results”, but that’s pretty generic.

I guess the “plugin” terminology was inherited from Flake8, but yeah, it’s not going to make much sense in the future where Ruff takes over the world to the point where people stop writing Flake8 plugins for us to clone and start contributing directly to Ruff. Usually “plugins” are things that can be separately installed and added at runtime (maybe we’ll have that in the future?).

**Comment by charliermarsh** [comment]
“When Ruff takes over the world” made my day 😄

**Comment by not-my-profile** [comment]
I like your suggestion of rule & violation since these words are commonly understood and the relationship between these concepts is clear: when you violate a rule you have a violation. Fun sidenote: Google claims to have >600k results for "violate a rule" but only 3 results for "violate a lint".

I don't really see the point of grouping rule implementations by their function signatures into files like `checks.rs` and `plugins.rs` ... I think it would make more sense to have one `.rs` file per rule, which would contain the function implementing the rule (no matter if it takes a Checker or returns a Check) and would also contain the definition of the new violation struct & `Violation` implementation (see #1685). Especially because the idea is to document the rules as docstrings on the structs, so it would be nice to have everything important for a rule (implementation, struct, struct impl and documentation) in one file.

**Comment by charliermarsh** [comment]
Yeah let's run with "rule" and "violation", I think.

I don't feel strongly about grouping by signature. I more like that there's some nomenclature that indicates whether the function is pure or mutates the checker. We could even have a `_mut` suffix or something. It's not a hard requirement for me though.

**Comment by not-my-profile** [comment]
We need a better name for the `Check` struct:

```rust
pub struct Check {
    pub kind: CheckKind,
    pub location: Location,
    pub end_location: Location,
    pub fix: Option<Fix>,
    pub parent: Option<Location>,
}
```

After #1685 each variant of the `CheckKind` enum will contain structs implementing the `Violation` trait, so `kind` should probably be renamed to `violation` ...

We notably cannot name `Check` violation as well since that we want both the now-called Check struct and the new trait to be usable in the same module without having to rename them. I am starting to think `Diag` (short for diagnostic) would be a good name for the `Check` struct.

So the idea would be: a Diagnostic consists of a Violation, the context where the violation occured and optionally a suggested Fix.

**Comment by not-my-profile** [comment]
Haha I guess I might just as well use this issue as a general place to raise terminology issues:

I don't think what we currently call "category" should be called category. IMO pyflakes, pycodestyle, isort, etc. aren't categories ... they are origins. I think a much better use for the term "category" is [what clippy calls categories](https://github.com/rust-lang/rust-clippy#clippy), namely categories such as all, correctness, suspicious, style, complexity, perf, pedantic or nursery.

**Comment by charliermarsh** [comment]
Does `CheckKind` then become `ViolationKind`? Or `DiagnosticViolation`?

E.g.:

```rs
pub struct Diagnostic {
    pub violation: DiagnosticViolation,
    pub location: Location,
    pub end_location: Location,
    pub fix: Option<Fix>,
    pub parent: Option<Location>,
}
```

Is that right?

**Comment by charliermarsh** [comment]
(The last struct to consider here is `Message`, which maybe we just leave as-is for now. )

**Comment by not-my-profile** [comment]
What's the difference between `Message` and the current `Check`? They look very similar.

**Comment by charliermarsh** [comment]
They're very similar. They exist because `Check` is generated in a single-file context. `Message` then adds the filename to `Check` (and extracts the source, if the user has that setting enabled).

**Comment by charliermarsh** [comment]
@not-my-profile - For now, if you start work on any of the TODOs from #1685, can you file a new issue, and I'll assign you? (Or just comment here if you prefer -- either is fine.)

**Comment by charliermarsh** [comment]
This has been improved enough that I'm gonna go ahead and close out for now (as of https://github.com/charliermarsh/ruff/pull/1760).

**Comment by charliermarsh** [comment]
Thanks to @not-my-profile for the push here and for a bunch of great work on the migration itself.

**Comment by not-my-profile** [comment]
You're very welcome :) Thank you for being so supportive!
